pkg/autoscaler/autoscaler.go

```diff
@@ -3,126 +3,208 @@ package autoscaler
 			}
-			as.metricsMap[resourceName][as.metricName] = newMetrics
+
+			resourcesMetricsMap[resourceName][metricName] = value
 		}
 	}
+
+	return resourcesMetricsMap, nil
 }
 
-func (as *Autoscaler) addMetricEntry(resourceName scaler_types.Resource, metricType string, entry metricEntry) {
-	if _, found := as.metricsMap[resourceName]; !found {
-		as.metricsMap[resourceName] = make(map[string][]metricEntry)
+func (as *Autoscaler) checkResourceToScale(resource scaler_types.Resource, resourcesMetricsMap map[string]map[string]int) bool {
+	if _, found := resourcesMetricsMap[resource.Name]; !found {
+		as.logger.DebugWith("Resource does not have metrics data yet, keeping up", "resourceName", resource.Name)
+		return false
+	}
+
+	for _, scaleResource := range resource.ScaleResources {
+		metricName := scaleResource.GetKubernetesMetricName()
+		value, found := resourcesMetricsMap[resource.Name][metricName]
+		if !found {
+			as.logger.DebugWith("One of the metrics is missing data, keeping up",
+				"resourceName", resource.Name,
+				"metricName", metricName)
+			return false
+		}
+
+		if value > scaleResource.Threshold {
+			as.logger.DebugWith("Metric value above threshold, keeping up",
+				"resourceName", resource.Name,
+				"metricName", metricName,
+				"threshold", scaleResource.Threshold,
+				"value", value)
+			return false
+		}
+
+		as.logger.DebugWith("Metric value below threshold",
+			"resourceName", resource.Name,
+			"metricName", metricName,
+			"threshold", scaleResource.Threshold,
+			"value", value)
 	}
-	as.metricsMap[resourceName][metricType] = append(as.metricsMap[resourceName][metricType], entry)
+
+	as.logger.DebugWith("All metric values below threshold, should scale to zero", "resourceName", resource.Name)
+	return true
 }
 
-func (as *Autoscaler) ReportMetric(metric metricEntry) error {
+func (as *Autoscaler) getBiggestWindow(resource scaler_types.Resource) time.Duration {
+	biggestWindow := 0 * time.Second
+	for _, scaleResource := range resource.ScaleResources {
+		if scaleResource.WindowSize > biggestWindow {
+			biggestWindow = scaleResource.WindowSize
+		}
+	}
+	return biggestWindow
+}
 
-	// don't block, try and fail fast
-	select {
-	case as.metricsChannel <- metric:
+func (as *Autoscaler) checkResourcesToScale(t time.Time) error {
+	activeResources, err := as.resourceScaler.GetResources()
+	if err != nil {
+		return errors.Wrap(err, "Failed to get resources")
+	}
+	if len(activeResources) == 0 {
+		as.logger.DebugWith("No active resources")
 		return nil
-	default:
-		as.logger.WarnWith("Failed to report metric",
-			"resourceName", metric.resourceName,
-			"MetricName", metric.metricName)
 	}
+	metricNames := as.getMetricNames(activeResources)
+	resourcesMetricsMap, err := as.getResourcesMetrics(metricNames)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get resources metrics")
+	}
+
+	for _, resource := range activeResources {
+		inScaleToZeroProcess, found := as.inScaleToZeroProcessMap[resource.Name]
```

Comment:
inScaleToZeroProcessMap is state... Why can't you get this per iteration?

Reply:
Talked offline - it's kind of a state but one that does not brake the scaler resistant to restarts.
If it will get restarted, this map will be empty, and new scale to zero process will occur, which is ok, since the process should be Idempotent, and it's good, since if the scaler pod restarted during the scale to zero process, the process will be stopped in the middle, that's why I don't want this state to be in any CRD, to prevent a state where it ignores a resource cause it sees that it is already scaling to zero, although it does not really in process (since it terminated)